ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -121,29 +146,30 @@ private void sendReply(DataStreamRequestByteBuf request, ChannelHandlerContext c
   private ChannelInboundHandler getServerHandler(){
     return new ChannelInboundHandlerAdapter(){
       @Override
-      public void channelRead(ChannelHandlerContext ctx, Object msg) {
+      public void channelRead(ChannelHandlerContext ctx, Object msg) throws IOException {
         final DataStreamRequestByteBuf request = (DataStreamRequestByteBuf)msg;
         final ByteBuf buf = request.slice();
         final boolean isHeader = request.getStreamOffset() == -1;
 
-        CompletableFuture<?>[] parallelWrites = new CompletableFuture<?>[clients.size() + 1];
+        CompletableFuture<?>[] parallelWrites = new CompletableFuture<?>[peers.size() + 1];
 
         final CompletableFuture<?> localWrites = isHeader?
                 streams.computeIfAbsent(request.getStreamId(), id -> getDataStreamFuture(buf))
                 : streams.get(request.getStreamId()).thenApply(stream -> writeTo(buf, stream));
         parallelWrites[0] = localWrites;
-        peersStreamOutput.putIfAbsent(request.getStreamId(), getDataStreamOutput());
 
           // do not need to forward header request
         if (isHeader) {
-          for (int i = 0; i < peersStreamOutput.get(request.getStreamId()).size(); i++) {
-            parallelWrites[i + 1] = peersStreamOutput.get(request.getStreamId()).get(i).getHeaderFuture();
+          final List<DataStreamOutput> outs = getDataStreamOutput();
+          peersStreamOutput.put(request.getStreamId(), outs);
+          for (int i = 0; i < outs.size(); i++) {
+            parallelWrites[i + 1] = outs.get(i).getHeaderFuture();
           }
         } else {
           // body
-          for (int i = 0; i < clients.size(); i++) {
-            parallelWrites[i + 1]  =
-              peersStreamOutput.get(request.getStreamId()).get(i).writeAsync(request.slice().nioBuffer());
+          final List<DataStreamOutput> outs = peersStreamOutput.get(request.getStreamId());
+          for (int i = 0; i < peers.size(); i++) {
```

Comment:
How about peers.size() -> outs.size() ?

Reply:
After refactoring, size() is not needed.  Will push a change soon.